examples/dataflow-bigquery-transpose/src/main/java/com/google/cloud/pso/common/PivotInputProvider.java

```diff
@@ -0,0 +1,246 @@
+      checkArgument(
+          valueFieldNames != null, "withValueFieldNames(valueFieldNames) called with null input");
+      checkArgument(
+          valueFieldNames.size() > 0,
+          "withValueFieldNames(valueFieldNames) called with empty input");
+      return setValueFieldNames(valueFieldNames);
+    }
+
+    /**
+     * Builds and returns a {@link PivotInputProvider} object.
+     *
+     * @return {@link PivotInputProvider} object.
+     */
+    public PivotInputProvider build() {
+      checkArgument(inputTableSchema() != null, "inputTableSchema is missing.");
+      checkArgument(keyFieldNames() != null, "keyFieldNames is missing.");
+      checkArgument(valueFieldNames() != null, "valueFieldNames is missing.");
+
+      // Validate whether the field names are present in the Schema
+      validateFields(keyFieldNames());
+      validateFields(pivotFieldNames());
+      validateFields(valueFieldNames());
+
+      // Validate whether the pivot field types are valid.
+      validatePivotFieldType();
+
+      Builder builder =
+          this.setKeyFieldSchema(extractSchema(keyFieldNames()))
+              .setPivotFieldSchema(extractSchema(pivotFieldNames()))
+              .setValueFieldSchema(extractSchema(valueFieldNames()));
+
+      Set<String> keyFieldNamesSet = Sets.newHashSet(keyFieldNames());
+      List<String> nonKeyFieldNames =
+          inputTableSchema()
+              .getFields()
+              .stream()
+              .map(Field::getName)
+              .filter(name -> !keyFieldNamesSet.contains(name))
+              .collect(Collectors.toList());
+
+      builder = builder.setNonKeySchema(extractSchema(nonKeyFieldNames));
+
+      return builder.autoBuild();
+    }
+
+    /** Validates that {@link List<String>} fieldnames is present in the {@link Schema}. */
+    private void validateFields(List<String> fieldNames) {
+      populateFieldNameToFieldMap();
+      Set<String> validFieldNames = fieldNameToFieldMap.keySet();
+
+      for (String fieldName : fieldNames) {
+        if (!validFieldNames.contains(fieldName)) {
+          throw new IllegalArgumentException(
+              "Invalid field name: " + fieldName + " in: " + inputTableSchema().toString());
+        }
+      }
+    }
+
+    /**
+     * Extract {@link Schema} from the input table {@link Schema} based on {@link List<String>}
+     * fieldNames.
+     */
+    private Schema extractSchema(List<String> fields) {
+      List<Field> extractedFields = Lists.newArrayListWithCapacity(fields.size());
+      populateFieldNameToFieldMap();
+
+      fields.forEach(field -> extractedFields.add(fieldNameToFieldMap.get(field)));
+
+      return Schema.of(extractedFields);
+    }
+
+    /** Validate that the pivot fields have the right {@link StandardSQLTypeName}. */
+    private void validatePivotFieldType() {
+      populateFieldNameToFieldMap();
+
+      for (String fieldName : pivotFieldNames()) {
+        if (!ALLOWED_PIVOT_TYPES.contains(
```

Comment:
This if-statement may be more readable if you write a method to evaluate the conditional logic and return the boolean result.